### PR TITLE
Add best checkpointer based on val_acc1

### PR DESCRIPTION
**Pull request by tsirif**

### CODE REVIEW

File: solo/utils/checkpointer.py [hunk 1]
```diff
@@ -135,6 +136,14 @@ def save(self, trainer: pl.Trainer):
                 os.remove(self.last_ckpt)
             self.last_ckpt = ckpt
 
+            current_val = float(trainer.callback_metrics['val_acc1'])
+            best_ckpt = self.path / 'best.ckpt'
```
Comment (lavoiems) [inline solo/utils/checkpointer.py:140]: I don't understand this line. What is the / operation for strings?

**Review by lavoiems — COMMENTED** [review]
They also get the accuracy from the momentum encoder which seems to be competitive with the accuracy from the online encoder.

Also, I am implementing the classifier on top of the y representation for byol.

**Comment by tsirif** [comment]
So wait are you suggesting to take the max out of all of them? in online fashion I think they only calculate from the 'online encoder'. Check the `validate_on_epoch` stuff inside `methods/base.py`.

File: solo/utils/checkpointer.py (hunk 1 above)
Comment (tsirif) [reply solo/utils/checkpointer.py:140]: It is not for strings, `/` is overloaded in the `Path` object of `pathlib` apparently. Check line `132`, that's what they do as well.

**Comment by lavoiems** [comment]
I think it depends on the model. E.g. BYOL is a BaseMomentumMethod which inherit from BaseModel and implements a classifier on top of the momentum network.

I am thinking about having an attribute that each model defines as the best over all the accuracy relevant to that model. We could define current_val as this score.

File: solo/utils/checkpointer.py (hunk 1 above)
Comment (lavoiems) [reply solo/utils/checkpointer.py:140]: Oh nice!

**Comment by tsirif** [comment]
So that value should be the one given to orion as well, I guess... even
though it might be the case that these values are extremely correlated

Στις Τετ, 22 Δεκ 2021, 20:40 ο χρήστης Sam ***@***.***>
έγραψε:

**Comment by lavoiems** [comment]
You are right. Lets merge for now and modify later.

**Pull request merged**